{"url": "https://api.github.com/repos/rust-lang/rust/issues/50301", "repository_url": "https://api.github.com/repos/rust-lang/rust", "labels_url": "https://api.github.com/repos/rust-lang/rust/issues/50301/labels{/name}", "comments_url": "https://api.github.com/repos/rust-lang/rust/issues/50301/comments", "events_url": "https://api.github.com/repos/rust-lang/rust/issues/50301/events", "html_url": "https://github.com/rust-lang/rust/issues/50301", "id": 318641116, "node_id": "MDU6SXNzdWUzMTg2NDExMTY=", "number": 50301, "title": "Program with HRTB rejected with inexplicable error message", "user": {"login": "ytausky", "id": 48800, "node_id": "MDQ6VXNlcjQ4ODAw", "avatar_url": "https://avatars.githubusercontent.com/u/48800?v=4", "gravatar_id": "", "url": "https://api.github.com/users/ytausky", "html_url": "https://github.com/ytausky", "followers_url": "https://api.github.com/users/ytausky/followers", "following_url": "https://api.github.com/users/ytausky/following{/other_user}", "gists_url": "https://api.github.com/users/ytausky/gists{/gist_id}", "starred_url": "https://api.github.com/users/ytausky/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/ytausky/subscriptions", "organizations_url": "https://api.github.com/users/ytausky/orgs", "repos_url": "https://api.github.com/users/ytausky/repos", "events_url": "https://api.github.com/users/ytausky/events{/privacy}", "received_events_url": "https://api.github.com/users/ytausky/received_events", "type": "User", "site_admin": false}, "labels": [{"id": 46741598, "node_id": "MDU6TGFiZWw0Njc0MTU5OA==", "url": "https://api.github.com/repos/rust-lang/rust/labels/E-needs-test", "name": "E-needs-test", "color": "02e10c", "default": false, "description": "Call for participation: writing correctness tests"}, {"id": 211668019, "node_id": "MDU6TGFiZWwyMTE2NjgwMTk=", "url": "https://api.github.com/repos/rust-lang/rust/labels/T-lang", "name": "T-lang", "color": "bfd4f2", "default": false, "description": "Relevant to the language team, which will review and decide on the PR/issue."}, {"id": 604454086, "node_id": "MDU6TGFiZWw2MDQ0NTQwODY=", "url": "https://api.github.com/repos/rust-lang/rust/labels/A-inference", "name": "A-inference", "color": "f7e101", "default": false, "description": "Area: Type inference"}, {"id": 650731663, "node_id": "MDU6TGFiZWw2NTA3MzE2NjM=", "url": "https://api.github.com/repos/rust-lang/rust/labels/C-bug", "name": "C-bug", "color": "f5f1fd", "default": false, "description": "Category: This is a bug."}], "state": "closed", "locked": false, "assignee": null, "assignees": [], "milestone": null, "comments": 2, "created_at": "2018-04-28T14:29:33Z", "updated_at": "2019-08-15T16:31:26Z", "closed_at": "2019-08-15T16:31:26Z", "author_association": "CONTRIBUTOR", "active_lock_reason": null, "body": "(Sorry about the non-informative title, I couldn't come up with a useful description.)\r\n\r\nAs discovered towards the end of [this forum thread](https://users.rust-lang.org/t/declaring-associated-item-whose-references-implement-intoiterator/17103/9?u=ytausky), the following code ([playground](https://play.rust-lang.org/?gist=9a1a8335e9b2ce64f468af5dc7bfaa45&version=stable&mode=debug)) fails to compile:\r\n```rust\r\ntrait Trait\r\nwhere\r\n    for<'a> &'a Self::IntoIter: IntoIterator<Item = u32>,\r\n{\r\n    type IntoIter;\r\n    fn get(&self) -> Self::IntoIter;\r\n}\r\n\r\nstruct Impl(Vec<u32>);\r\n\r\nimpl Trait for Impl {\r\n    type IntoIter = ImplIntoIter;\r\n    fn get(&self) -> Self::IntoIter {\r\n        ImplIntoIter(self.0.clone())\r\n    }\r\n}\r\n\r\nstruct ImplIntoIter(Vec<u32>);\r\n\r\nimpl<'a> IntoIterator for &'a ImplIntoIter {\r\n    type Item = <Self::IntoIter as Iterator>::Item;\r\n    type IntoIter = std::iter::Cloned<std::slice::Iter<'a, u32>>;\r\n    fn into_iter(self) -> Self::IntoIter {\r\n        (&self.0).into_iter().cloned()\r\n    }\r\n}\r\n```\r\nI get the following error message:\r\n```\r\nerror[E0271]: type mismatch resolving `for<'a> <std::slice::Iter<'a, u32> as std::iter::Iterator>::Item == &u32`\r\n  --> src/lib.rs:13:6\r\n   |\r\n13 | impl Trait for Impl {\r\n   |      ^^^^^ expected bound lifetime parameter 'a, found concrete lifetime\r\n   |\r\n   = note: required because of the requirements on the impl of `std::iter::Iterator` for `std::iter::Cloned<std::slice::Iter<'_, u32>>`\r\n```\r\nChanging the definition of `Item` to `u32` makes the program compile, but it seems like it should be accepted as it is.", "closed_by": {"login": "bors", "id": 3372342, "node_id": "MDQ6VXNlcjMzNzIzNDI=", "avatar_url": "https://avatars.githubusercontent.com/u/3372342?v=4", "gravatar_id": "", "url": "https://api.github.com/users/bors", "html_url": "https://github.com/bors", "followers_url": "https://api.github.com/users/bors/followers", "following_url": "https://api.github.com/users/bors/following{/other_user}", "gists_url": "https://api.github.com/users/bors/gists{/gist_id}", "starred_url": "https://api.github.com/users/bors/starred{/owner}{/repo}", "subscriptions_url": "https://api.github.com/users/bors/subscriptions", "organizations_url": "https://api.github.com/users/bors/orgs", "repos_url": "https://api.github.com/users/bors/repos", "events_url": "https://api.github.com/users/bors/events{/privacy}", "received_events_url": "https://api.github.com/users/bors/received_events", "type": "User", "site_admin": false}, "reactions": {"url": "https://api.github.com/repos/rust-lang/rust/issues/50301/reactions", "total_count": 0, "+1": 0, "-1": 0, "laugh": 0, "hooray": 0, "confused": 0, "heart": 0, "rocket": 0, "eyes": 0}, "timeline_url": "https://api.github.com/repos/rust-lang/rust/issues/50301/timeline", "performed_via_github_app": null, "state_reason": "completed"}